### PR TITLE
document: allow record duplication

### PR DESCRIPTION
**Pull request by Garfield-fr**
This feature allows you to duplicate a document by clicking on the duplicate button in the detailed view.

Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>

## How to test?

- Select a document and click "duplicate" button.
- Modify data and save.

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Extracted translations?

### CODE REVIEW

**Review by iGormilhit — CHANGES_REQUESTED** [review]
Commit message title proposition: `document: allow record duplication`

**Review by zannkukai — APPROVED** [review]

**Comment by Garfield-fr** [comment]
> Commit message title proposition: `document: allow record duplication`

done

File: projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.html [hunk 1]
```diff
@@ -19,8 +19,13 @@
   <a routerLink="/records/documents/new" [queryParams]="{source: source, pid: pid}" class="btn btn-sm btn-primary"
     translate>Import</a>
 </div>
+<!-- duplicate button -->
+<div *ngIf="record && record.metadata && record.metadata.pid" class="float-right mt-2 mb-4 btn-duplicate">
+  <a routerLink="/records/documents/duplicate" [queryParams]="{type: 'documents', pid: record.metadata.pid}" class="btn btn-sm btn-primary"
```
Comment (AoNoOokami) [inline projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.html:24]: I think the right syntax is: `[routerLink]="['/records', 'documents', 'duplicate']">`
See line 108.
And it's the occasion to correct line 19 too.

**Review by AoNoOokami — CHANGES_REQUESTED** [review]

File: projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.html (hunk 1 above)
Comment (Garfield-fr) [reply projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.html]: done

**Review by iGormilhit — APPROVED** [review]

File: projects/admin/src/app/record/custom-editor/document-editor/document-editor.component.ts [hunk 2]
```diff
@@ -73,11 +102,14 @@ export class DocumentEditorComponent {
   loadingChanged(value: boolean) {
     if (value === false) {
       combineLatest([this._route.params, this._route.queryParams])
-        .subscribe(([params, queryParams]) => {
-          if (queryParams.source != null && queryParams.pid != null) {
-            this.importFromExternalSource(queryParams.source, queryParams.pid);
-          }
-        });
+      .subscribe(([params, queryParams]) => {
+        if (queryParams.source != null && queryParams.pid != null) {
```
Comment (jma) [inline projects/admin/src/app/record/custom-editor/document-editor/document-editor.component.ts:106]: you can factorize: `queryParams.pid != null` in a separate `if` statement.

File: projects/admin/src/app/service/editor.service.ts [hunk 3]
```diff
@@ -1,33 +1,58 @@
...
   ) { }
 
+  /**
+   * Get record from rero-ils
+   * @param type - Type of resource
+   * @param pid - pid of resource
+   * @return Observable or error
+   */
+  getRecord(type: string, pid: string): Observable<any | Error> {
+    return this._recordService.getRecord(type, pid);
```
Comment (jma) [inline projects/admin/src/app/service/editor.service.ts:45]: this seems useless.

File: projects/admin/src/app/record/custom-editor/document-editor/document-editor.component.ts [hunk 4]
```diff
@@ -48,14 +55,36 @@ export class DocumentEditorComponent {
...
+    );
+  }
+
+  /**
+   * Get record by type and pid
+   * @param type string - resource type
+   * @param pid - resource pid
+   */
+  duplicateRecord(type: string, pid: string) {
+    this._editorService.getRecord(type, pid).subscribe(
```
Comment (jma) [inline projects/admin/src/app/record/custom-editor/document-editor/document-editor.component.ts:78]: why not using the recordService directly?

**Review by jma — APPROVED** [review]

File: projects/admin/src/app/service/editor.service.ts (hunk 3 above)
Comment (Garfield-fr) [reply projects/admin/src/app/service/editor.service.ts]: done

File: projects/admin/src/app/record/custom-editor/document-editor/document-editor.component.ts (hunk 4 above)
Comment (Garfield-fr) [reply projects/admin/src/app/record/custom-editor/document-editor/document-editor.component.ts]: done

File: projects/admin/src/app/record/custom-editor/document-editor/document-editor.component.ts (hunk 2 above)
Comment (Garfield-fr) [reply projects/admin/src/app/record/custom-editor/document-editor/document-editor.component.ts]: Changed

**Pull request merged**